datasets/utils/secret.py

```diff
@@ -0,0 +1,210 @@
+import base64
+import json
+import logging
+import os
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Dict, Union
+
+import boto3
+from kubernetes import client, config
+from tenacity import (
+    Retrying,
+    retry_if_exception,
+    stop_after_attempt,
+    wait_fixed,
+)
+
+
+logger = logging.getLogger(__name__)
+
+# Ensure we return a non-None value
+# Prefer to throw exception rather than returning None if secret can't be found
+secret_return_type = Union[str, Dict[str, str]]
+
+# Global secret cache to prevent redundant
+# secret retrieval calls to external system if possible
+secret_cache = defaultdict(dict)
+
+
+@dataclass
+class Secret:
+    """
+    Dataclass interface for secret retrieval.
+
+    Need to provide exact one of the following secret source variable:
+        cluster_secret_name: name of cluster secret.
+            secret value to be pulled from cluster secret
```

Comment:
`k8s_cluster_secret_name` ?

Reply:
Renaming to `kubernetes_secret_name` to also clear the confusion that this is not a cluster-wide secret